trie/triedb/pathdb/disklayer.go

```diff
@@ -0,0 +1,259 @@
+)
+
+// diskLayer is a low level persistent layer built on top of a key-value store.
+type diskLayer struct {
+	rootHash common.Hash  // Immutable, root hash of the base layer
+	id       uint64       // Immutable, corresponding state id
+	db       *Database    // Path-based trie database
+	buffer   *nodebuffer  // Node buffer to aggregate writes.
+	stale    bool         // Signals that the layer became stale (state progressed)
+	lock     sync.RWMutex // Lock used to protect stale flag
+}
+
+// newDiskLayer creates a new disk layer based on the passing arguments.
+func newDiskLayer(root common.Hash, id uint64, db *Database, buffer *nodebuffer) *diskLayer {
+	return &diskLayer{
+		rootHash: root,
+		id:       id,
+		db:       db,
+		buffer:   buffer,
+	}
+}
+
+// root returns root hash of corresponding state.
+func (dl *diskLayer) root() common.Hash {
+	return dl.rootHash
+}
+
+// parent always returns nil as there's no layer below the disk.
+func (dl *diskLayer) parent() layer {
+	return nil
+}
+
+// stateID returns the state id of disk layer.
+func (dl *diskLayer) stateID() uint64 {
+	return dl.id
+}
+
+// isStale return whether this layer has become stale (was flattened across) or if
+// it's still live.
+func (dl *diskLayer) isStale() bool {
+	dl.lock.RLock()
+	defer dl.lock.RUnlock()
+
+	return dl.stale
+}
+
+// markStale sets the stale flag as true.
+func (dl *diskLayer) markStale() {
+	dl.lock.Lock()
+	defer dl.lock.Unlock()
+
+	if dl.stale {
+		panic("triedb disk layer is stale") // we've committed into the same base from two children, boom
+	}
+	dl.stale = true
+}
+
+// Node retrieves the trie node with the provided node info. No error will be
+// returned if the node is not found.
+func (dl *diskLayer) Node(owner common.Hash, path []byte, hash common.Hash) ([]byte, error) {
+	dl.lock.RLock()
+	defer dl.lock.RUnlock()
+
+	if dl.stale {
+		return nil, errSnapshotStale
+	}
+	// Try to retrieve the trie node from the not-yet-written
+	// node buffer first. Note the buffer is lock free since
+	// it's impossible to mutate the buffer before tagging the
+	// layer as stale.
+	n, err := dl.buffer.node(owner, path, hash)
+	if err != nil {
+		return nil, err
+	}
+	if n != nil {
+		dirtyHitMeter.Mark(1)
+		dirtyReadMeter.Mark(int64(len(n.Blob)))
+		return n.Blob, nil
+	}
+	dirtyMissMeter.Mark(1)
+
+	// Try to retrieve the trie node from the clean memory cache
+	if dl.db.cleans != nil {
+		if blob := dl.db.cleans.Get(nil, hash.Bytes()); len(blob) > 0 {
+			cleanHitMeter.Mark(1)
+			cleanReadMeter.Mark(int64(len(blob)))
+			return blob, nil
+		}
+		cleanMissMeter.Mark(1)
+	}
+	// Try to retrieve the trie node from the disk.
+	var (
+		nBlob []byte
+		nHash common.Hash
+	)
+	if owner == (common.Hash{}) {
+		nBlob, nHash = rawdb.ReadAccountTrieNode(dl.db.diskdb, path)
+	} else {
+		nBlob, nHash = rawdb.ReadStorageTrieNode(dl.db.diskdb, owner, path)
+	}
+	if nHash != hash {
+		return nil, &UnexpectedNodeError{
```

Comment:
The error message already contains where the error is raised. See the type field.

```go
func (err *UnexpectedNodeError) Error() string {
	return fmt.Sprintf("%s: unexpected node %x!=%x(%x %v)", err.typ, err.expected, err.hash, err.owner, err.path)
}
```

Reply:
By "where" I meant "when the node is not found". I.e. instead of saying "unexpected node, hash == empty hash (you better recognize it)", it would be "missing node". If we know we must have a node but don't have it, we could be a lot noisier than just an unexpected node error return.

Reply:
Gotcha, make sense.